### DISCUSSION TITLE
Enable zero downtime/graceful shutdown

### DISCUSSION THREAD

**Discussion by tmeijn** [npalm/terraform-aws-gitlab-runner#1067] [Ideas]
Hey @kayman-mk I've been thinking a bit on how to implement graceful shutdown and therefore having the ability to do zero downtime deployments. This is just high level based on some research I did, nothing actually tested yet, want to get your thoughts:

1. ~~We run a script on the instance in the background that polls whether the instance is in the terminating state, ref: https://stackoverflow.com/questions/35914004/how-to-detect-state-of-aws-instance-from-inside-of-itself~~
    * scratch that, we can actually use the [instance metadata api to monitor the target state](https://docs.aws.amazon.com/autoscaling/ec2/userguide/retrieving-target-lifecycle-state-through-imds.html):
    
![image](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/17970041/76024e22-c88b-4ffc-89d6-96edc6c5afbc)

![image](https://github.com/cattle-ops/terraform-aws-gitlab-runner/assets/17970041/7b5d8e56-8acb-4360-aa24-776c86d4972c)

2. Mark the service as non-restartable: https://unix.stackexchange.com/questions/689011/how-to-prevent-restarting-of-systemd-services
3. Once this state is detected, send a [SIGQUIT signal](https://docs.gitlab.com/runner/commands/#signals) to the Runner process. According to the docs this should do a graceful shutdown.
4. Monitor the existence of the process..., if it's still there, [send heartbeat for the lifecycle](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/autoscaling/record-lifecycle-action-heartbeat.html)
5. Once Runner process is no longer there, send [complete-lifecycle-action](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/autoscaling/complete-lifecycle-action.html)
6. Profit?

The "Cleanup Lambda" would have to be triggered on actual termination instead of the terminate-wait hook.

Please let me know what you think! 😄

**Comment by kayman-mk** [comment]
Just linking #1029 and #631. Now on my list. We should push this forward.

**Reply by kayman-mk** [reply]
Since https://github.com/cattle-ops/terraform-aws-gitlab-runner/pull/1117 (May 2024) this has been solved. The old Runners are killed as soon as they finished the current jobs. The new Runners are spun up in parallel.
